### PR TITLE
AvatarRoot API: Non-VRChat avatar support in VRCSDK projects

### LINKED ISSUE #68: Improve RuntimeUtil.IsAvatarRoot()
https://github.com/bdunderscore/ndmf/blob/2d0357f11f2dda7f6894d90b105f6b16fc73675c/Runtime/RuntimeUtil.cs#L95-L105

## as-is

- VRCSDK project: true for VRC avatars, false for non-VRC avatars
- non-VRCSDK project: true for VRC and non-VRC avatars

## to-be

- VRCSDK projects should be able to handle non-VRC avatars, but not lenient enough to treat all Animators as avatars
- Something like...?
  - If any avatar platform is detected (for example, VRCSDK and/or uniVRM), iterate through all platform specific checks
  - Else, fallback to Animator heuristic (current non-VRC implementation)

### PR DESCRIPTION
**Pull request by kaikoga**
Resolves #68, with avatar root detection logic described in the issue.

This is a potentially breaking change for VRCSDK + UniVRM environment. Current Non-VRChat avatar support (implemented in #34) is meant to not break existing NDMF plugins, which targets VRChat avatars only, in all projects including VRCSDK + UniVRM setup. This PR is the breaking part.

Explanation: VRM avatars were not AvatarRoots but would be. NDMF plugins expecting VRChat avatars may try to process VRM avatars, which would fail because VRCAvatarDescriptor does not exist. VRCSDK + UniVRM environment can be created using legacy UniVRM (0.99.4). [info](https://vrm-c.github.io/UniVRM/)

User mitigation: Remove all VRM avatars, or all `VRMMeta` / `Vrm10Instance` components, when using VRChat only NDMF plugins in hybrid projects.

### CODE REVIEW

**Comment by bdunderscore** [comment]
Hmm. To resolve the compatibility issue, I think I'd prefer to have plugins declare what environments they're compatible with (with "vrchat" being the default). Let me noodle on this a bit and add that API first, before merging this.

**Comment by bdunderscore** [comment]
This will also need tests in a UniVRM environment... I'll have to look into auto-importing unitypackages.

**Comment by anatawa12** [comment]
I think we should use upm version of univrm for testing since this pr expects upm version of univrm, not unitypackage one.

**Comment by kaikoga** [comment]
If ndmf is going to be aware of environments, perhaps we can warn if VRM version defines are not defined but installation is detected through `AppDomain.CurrentDomain.GetAssemblies()` (this should indicate a unitypackage installation)

**Comment by bdunderscore** [comment]
Sorry for the lack of replies - I haven't been able to do much development on NDMF/MA over the past few weeks. I'll hopefully be able to spend some more time on it in December.

**Comment by kaikoga** [comment]
Thoughts for this issue:

Theoretically, there may be three types of avatar setup when we want to create an avatar for VRC and VRM.
Which strategy should we support (or recommend)?

### Setup as separate avatars

- VRCAvatar (VRCDescriptor)
- VRMAvatar (VRMMeta)

Pros: Currently available.
Cons: When VRC avatar (or MA outfit) is modified, manual sync to VRM avatar (or MA outfit) by hand.

### Conversion plugin

- VRCAvatar (VRCDescriptor)
  - → some ndmf plugin converts VRC components into VRM counterparts

Pros: Easy.
Cons: Requires development of conversion plugin, which may not be accurate enough or well customizable.

### Universal avatar

- VRCAvatar (VRCDescriptor) (VRMMeta)
  - → some ndmf plugin strips VRM components in VRC builds, vice versa.

Pros: Possible. Proof of concept ( [docs_ja](https://docs.kaikoga.net/ativ/ndmf/ativ_select_platform) [component](https://github.com/kaikoga/AvatarTinkerVista-Unity/blob/4b1e015676a8e98e02254fb21c2337bb070debd2/Runtime/Silksprite/AvatarTinkerVista/Ndmf/AtivSelectPlatform.cs) [ndmf pass](https://github.com/kaikoga/AvatarTinkerVista-Unity/blob/4b1e015676a8e98e02254fb21c2337bb070debd2/Editor/Silksprite/AvatarTinkerVista/Ndmf/Passes/DeleteComponentsPass.cs) )
Cons: Somewhat bloated avatar.

File: Runtime/RuntimeUtil.cs [hunk 1]
```diff
@@ -95,8 +101,17 @@ public static string AvatarRootPath(GameObject child)
         public static bool IsAvatarRoot(Transform target)
```
Comment (bdunderscore) [inline Runtime/RuntimeUtil.cs:101]: こちらの要更新？（仮実装です）
https://github.com/bdunderscore/ndmf/blob/23fa9c4cfd9c5b7a6c9173107b0e264fb424abb2/Editor/PreviewSystem/ComputeContext/SingleObjectQueries.cs#L33-L44

Comment (kaikoga) [reply Runtime/RuntimeUtil.cs:101]: そうかもしれません・・・
NDMF 1.5.0 リリース後に作業を再開する予定でしたが、この機会に全体を見直しておきます。

Comment (bdunderscore) [reply Runtime/RuntimeUtil.cs:101]: 了解です。

File: Runtime/RuntimeUtil.cs [hunk 2]
```diff
@@ -94,14 +100,23 @@ public static string AvatarRootPath(GameObject child)
         /// <returns></returns>
         public static bool IsAvatarRoot(Transform target)
         {
+            // First, look for platform specific avatar descriptors
+            // TODO: ignore nested avatar descriptors?
```
Comment (kaikoga) [inline Runtime/RuntimeUtil.cs:104]: たぶん以下の扱いをちゃんと考えないといけない

- VRCAvatarDescriptor の子に Vrm10Instance
  - VRCAvatarDescriptor をアバターとみなすべき
- Vrm10Instance の子に VRCAvatarDescriptor
  - Vrm10Instance をアバターとみなすべき
- Animator の子に VRCAvatarDescriptor / Vrm10Instance
  - バニラの Animator はアバターではないかもしれない
    - VRChatプロジェクトだと VRCAvatarDescriptor がアバターっぽく見える
    - Resoniteアバターがメインターゲットになるなら Animator がアバターになるべきケースもある
      - その時はたぶん VRCAvatarDescriptor を外してもらった方が良い
    - FindAvatarsInScene() みたいに全てのアバターかもしれない Animator をリストアップするとエラー UI などが不便になるかもしれない
    - FindAvatarRoots() は少し賢い
  - Animator が実は CVRAvatar だったケース
    - CVRAvatar をアバターとみなしたいけど、上との区別が無理
    - VRCAvatarDescriptor を外すか、 ndmf が CVRAvatar に対応するか、 AvatarRoot マーカーコンポーネントを定義してつけてもらうか・・・

Comment (kaikoga) [reply Runtime/RuntimeUtil.cs:104]: 考えました #432

**Comment by kaikoga** [comment]
I have changed my mind, and would propose a more platform agnostic avatar root definition.

See #465.

File: Runtime/RuntimeUtil.cs (hunk 2 above)
Comment (kaikoga) [reply Runtime/RuntimeUtil.cs:104]: > Animator の子に VRCAvatarDescriptor / Vrm10Instance

「アバターの外側に Animator をつけたい」というケースがない気がしました（外側の Animator はアバターの Humanoid ボーンや Animator パラメータを触ることができないので）

**Pull request closed without merging**